### ISSUE TITLE
Synchronize enabled Mosaic with runtime input dimensions

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#65]
## Phase

Implementation authorized for the isolated runtime-dimension/Mosaic state defect proven by validation Issue #53.

This Issue does not authorize head redesign, sampler redesign, transform redesign, export work, training-policy changes, CI work, or model-quality experiments.

## Baseline

Use current `origin/main`:

`7d1337bf2f7dfbc94c4688cfa7d9868d400295be`

## Proven defect

Issue #53 demonstrated on the maintained HBB and OBB training data paths that runtime dimension changes are propagated correctly through the sampler, final preprocessing, target scaling, rectangular head grids, and decode, **except while Mosaic remains enabled**.

The causal state mismatch is:

- `DataLoader.change_input_dim(...)` schedules the new HxW on the batch sampler;
- `YoloBatchSampler` emits the new dimension tuple;
- `Dataset.resize_getitem` transiently exposes that active dimension through the outer Mosaic wrapper's `self.input_dim`;
- but the enabled Mosaic branches in both `MosaicDetection` and `MosaicDetectionOBB` use `self._dataset.input_dim` instead of the active wrapper dimension;
- the underlying dataset dimension is only explicitly synchronized in the no-Mosaic branch.

Therefore an enabled-Mosaic transition such as `416x416 -> 512x512` keeps Mosaic placement/canvas/random-perspective in the old 416 coordinate space while the final preprocessor/output tensor uses 512. Rectangular transitions reproduce the same defect.

Validation #53 proved:

- stale enabled-Mosaic canvas/RP for both HBB and OBB;
- pure HBB/OBB preprocessing is correct;
- classes/angles/field ordering are correct;
- HBB/OBB head grid regeneration and rectangular decode are correct;
- closing Mosaic avoids the stale path;
- no implementation was performed during validation.

## Required behavior

When Mosaic is enabled, the Mosaic canvas, placement geometry, resize scale, border, random-perspective coordinate space, optional MixUp input dimension, and final preprocessing must all use the **same active runtime HxW carried by the current sampled batch**.

Runtime dimension changes must work for square and rectangular shapes and must remain consistent across repeated changes.

The no-Mosaic path and existing Mosaic augmentation policy must remain unchanged apart from correcting the stale dimension source.

## Preferred causal boundary

Keep production changes local to:

- `yolox/data/datasets/mosaicdetection.py`
- `yolox/data/datasets/mosaicdetection_obb.py`

The active outer wrapper dimension (`self.input_dim`) is already provided by `Dataset.resize_getitem` during the decorated call and is the expected source of truth for the currently sampled batch.

Prefer using that active wrapper dimension directly in the enabled Mosaic path rather than redesigning sampler or head state.

Do not mutate/synchronize the underlying dataset `_input_dim` more broadly unless a focused regression demonstrates that such mutation is required for correctness. Avoid widening state lifetime unnecessarily.

## Required development sequence

1. Preflight:
   - confirm worktree clean;
   - fetch origin;
   - switch to `main`;
   - run `git pull --ff-only`;
   - verify local HEAD == origin/main == `7d1337bf2f7dfbc94c4688cfa7d9868d400295be`;
   - only then create a dedicated implementation branch.
2. Add focused regressions that fail on the baseline.
3. Implement the smallest causal HBB+OBB Mosaic dimension correction.
4. Prove focused regressions pass.
5. Run broader maintained local tests.
6. Commit with `Refs #65`.
7. Open a PR referencing #65.
8. Post a factual implementation handoff to #65.
9. STOP for orchestrator review.

If preflight fails or main has moved unexpectedly, STOP and report. Never discard local work.

## Required failing regression matrix

Exercise the actual maintained `DataLoader.change_input_dim` -> `YoloBatchSampler` -> `Dataset.resize_getitem` -> Mosaic wrapper path.

For both HBB and OBB cover at minimum:

### Enabled Mosaic dimension transitions

- `416x416 -> 512x512`
- `416x416 -> 320x512`
- `416x416 -> 512x320`
- `416x416 -> 384x640`
- `416x416 -> 640x384`

For each transition assert that the second/current batch uses the requested active dimension for:

- source-image Mosaic resize scale;
- Mosaic canvas shape `2H x 2W`;
- Mosaic center/placement coordinate space;
- random-perspective border `[-H//2, -W//2]`;
- random-perspective output HxW;
- final preprocessing HxW;
- emitted tensor HxW;
- transformed target coordinate system.

Use deterministic synthetic geometry and freeze random values only where required for reproducible expected values.

### Repeated transitions

At minimum:

`416x416 -> 384x640 -> 640x384 -> 512x512`

using the same wrapper/loader lifecycle where maintained semantics permit it.

Verify there is no stale canvas/RP dimension from the preceding batch.

### Mosaic phase boundary preservation

Cover:

- dimension change while Mosaic enabled;
- `close_mosaic()` then dimension change;
- dimension change scheduled, then `close_mosaic()` before consumption.

The already-correct no-Mosaic behavior must remain correct.

### Optional MixUp

Do not enable or redesign MixUp policy solely for this Issue.

However, inspect that if the enabled Mosaic path passes `self.input_dim` into existing MixUp, it continues to use the same active current HxW. A lightweight deterministic check is sufficient if MixUp can be exercised cheaply.

## Preservation requirements

Do not change:

- `DataLoader.change_input_dim` semantics;
- `YoloBatchSampler` pending-dimension semantics;
- positive-worker prefetch behavior;
- `Dataset.resize_getitem` API/lifetime semantics;
- HBB/OBB head grid/decode code;
- TrainTransform/TrainTransformOBB geometry;
- Mosaic center/randomization policy;
- random-perspective policy;
- MixUp enable/disable policy;
- target angle convention;
- training recipe/configuration;
- export/ONNX behavior;
- repository layout.

No optimizer step or training is required.
No GitHub CI work.

## Test quality requirements

- Do not test only a detached helper.
- The focused regression must pass through the maintained decorated Mosaic `__getitem__` path with runtime dimension tuples.
- Before-fix evidence must demonstrate stale internal canvas/RP geometry while the requested/final dimension differs.
- After-fix evidence must demonstrate all internal and final coordinate spaces agree with the current sampled HxW.
- Cover HBB and OBB separately; do not assume one proves the other.

## Broader validation

Run the relevant maintained augmentation/target/grid tests plus the broader maintained local suite where practical.

Do not use GitHub CI.

## Handoff

Report:

- exact preflight baseline;
- branch/commit/head SHA;
- PR number/URL;
- changed files;
- failing regression evidence before the fix;
- exact production diff summary;
- HBB transition matrix after fix;
- OBB transition matrix after fix;
- Mosaic canvas/RP/current-dimension evidence;
- phase-boundary preservation evidence;
- MixUp compatibility check if exercised;
- broader local test results;
- FACT / INFERENCE / UNVERIFIED.

Do not merge your own PR.
Do not close #65.
STOP for orchestrator review.

**Comment by shapovalovei** [comment]
## Implementation handoff

Implementation is complete on the dedicated branch; the Issue remains open for orchestrator review.

### Identity and scope

- Repository: `shapovalovei/YOLOX-OBB`
- Required preflight: clean worktree; `git fetch origin`; `git switch main`; `git pull --ff-only` already up to date; local `HEAD == origin/main == 7d1337bf2f7dfbc94c4688cfa7d9868d400295be`.
- Branch: `fix/65-mosaic-runtime-input-dim`
- Commit: `0ee6ef9c49264ac7439801b0cb1eb3a4fc9141bb`
- Draft PR: [#66](https://github.com/shapovalovei/YOLOX-OBB/pull/66)
- Changed files: `yolox/data/datasets/mosaicdetection.py`, `yolox/data/datasets/mosaicdetection_obb.py`, and `tests/test_mosaic_runtime_dimensions.py`.
- Production changes are limited to the two requested Mosaic wrappers. No DataLoader, sampler, dataset-wrapper API, head/decode, transform, policy, recipe, export, CI, training, or generated artifact changes.

### Baseline failing evidence

Before the production edit, the focused maintained-path regression run at the required baseline produced `12 failed, 8 passed in 7.45s`. The 10 square/rectangular enabled-Mosaic matrix subcases and both repeated-transition cases observed stale internal RP input `(832, 832)` from the original `416x416` state instead of the requested current canvas `(2H, 2W)`. The phase-boundary, no-Mosaic, and MixUp-forwarding checks passed on baseline, isolating the defect to enabled-Mosaic dimension selection.

Command:

`/Users/eugene/creditCardReader/venv/bin/python -m pytest -q tests/test_mosaic_runtime_dimensions.py`

### Exact production correction

The enabled branch now assigns `input_dim = self.input_dim` in both `MosaicDetection.__getitem__` and `MosaicDetectionOBB.__getitem__`. This uses the transient active dimension supplied by `Dataset.resize_getitem` for the current sampled tuple. No synchronization or persistence of `self._dataset._input_dim` was added; the no-Mosaic branch is unchanged.

### HBB post-fix transition matrix

The maintained path was exercised as `DataLoader.change_input_dim` -> `YoloBatchSampler` -> `Dataset.resize_getitem` -> decorated `MosaicDetection.__getitem__`, with deterministic synthetic `240x360` source geometry. Source resize is reported HxW; target center is the first synthetic HBB target after Mosaic/RP and before final transform.

| transition | source resize | Mosaic canvas | RP border | RP output / preproc / tensor | first target center |
|---|---:|---:|---:|---:|---:|
| `416x416 -> 512x512` | `341x512` | `1024x1024` | `[-256,-256]` | `512x512 / 512x512 / 512x512` | `(170.6667,171.0000)` |
| `416x416 -> 320x512` | `320x480` | `640x1024` | `[-160,-256]` | `320x512 / 320x512 / 320x512` | `(176.0000,80.0000)` |
| `416x416 -> 512x320` | `213x320` | `1024x640` | `[-256,-160]` | `512x320 / 512x320 / 512x320` | `(106.6667,203.0000)` |
| `416x416 -> 384x640` | `384x576` | `768x1280` | `[-192,-320]` | `384x640 / 384x640 / 384x640` | `(224.0000,96.0000)` |
| `416x416 -> 640x384` | `256x384` | `1280x768` | `[-320,-192]` | `640x384 / 640x384 / 640x384` | `(128.0000,256.0000)` |

### OBB post-fix transition matrix

The same maintained path was exercised independently through `MosaicDetectionOBB`. Every dimension row matched the HBB dimension evidence above: source resize, `2H x 2W` canvas, `[-H//2,-W//2]` border, HxW RP output, HxW preprocessor input, and emitted tensor HxW. First-target centers were `(170.6667,171.0000)`, `(176.0000,80.0000)`, `(106.6667,203.0000)`, `(224.0000,96.0000)`, and `(128.0000,256.0000)` respectively; OBB angle `25.0` degrees and class `3.0` were preserved in every row.

### Repeated and phase-boundary evidence

- Enabled Mosaic same-loader sequence `416x416 -> 384x640 -> 640x384 -> 512x512` passed independently for HBB and OBB. Each consumed batch used its own `2H x 2W` canvas and HxW RP/preproc/tensor dimensions; no preceding canvas/RP size persisted.
- `close_mosaic()` followed by scheduling `512x320` produced a no-Mosaic HBB and OBB batch at `512x320`; no RP call occurred and the underlying dataset dimension became `512x320`.
- Scheduling `384x640` first, then calling `close_mosaic()` before consumption, produced a no-Mosaic HBB and OBB batch at `384x640`; no RP call occurred and the underlying dataset dimension became `384x640`.
- Starting with Mosaic disabled and changing to `512x320` preserved no-Mosaic behavior. The emitted target center scaled from `(300,180)` to `(266.6667,160.0000)`; HBB class and OBB angle/class fields were preserved.
- Lightweight MixUp compatibility spies received `[416x416, 384x640]` for both HBB and OBB. MixUp policy was not enabled or redesigned in production; the actual MixUp algorithm was not otherwise changed.
- After enabled-Mosaic calls, the outer decorator state was cleaned (`_input_dim` absent), the wrapper default dimension remained `416x416`, and the underlying dataset remained `416x416`; no broad state synchronization was introduced.

### Validation

- Focused: `/Users/eugene/creditCardReader/venv/bin/python -m pytest -q tests/test_mosaic_runtime_dimensions.py` -> `10 passed, 10 subtests passed`.
- Broader maintained local suite: `/Users/eugene/creditCardReader/venv/bin/python -m unittest discover -s tests -p 'test_*.py' -v` -> `Ran 151 tests in 23.666s; OK`.\n- `git diff --check` passed. Python compilation passed for the two production modules and the new regression module.\n- No training, optimizer step, model-quality experiment, GitHub CI, export, or generated artifact work was run.\n\n### FACT / INFERENCE / UNVERIFIED\n\n**FACT**\n\n- The exact required baseline was verified before branching.\n- The baseline regression reproduced stale `416x416` enabled-Mosaic internal geometry for both HBB and OBB while final preprocessing/tensor dimensions followed the requested runtime tuple.\n- The two-line production correction makes all requested square/rectangular internal and final dimensions agree in deterministic maintained-path tests, preserves OBB angle/class fields, preserves no-Mosaic behavior, and forwards the active dimension to existing MixUp.\n- The broader local suite passed: 151 tests.\n\n**INFERENCE**\n\n- The smallest causal correction is using the active outer wrapper `self.input_dim` in each enabled-Mosaic branch. The proven sampler, decorator, transform, head, and no-Mosaic contracts require no changes.\n\n**UNVERIFIED**\n\n- No new positive-worker prefetch stress test was run in this implementation phase; #53\x27s prior bounded worker smoke remains the relevant evidence for unchanged sampler/prefetch behavior.\n- Distributed behavior, full forward/loss behavior, real datasets, training/model quality, export/ONNX behavior, and the full MixUp transformation were not evaluated.\n\nIntentionally unchanged: DataLoader and YoloBatchSampler semantics, Dataset.resize_getitem lifetime, HBB/OBB heads and decode, TrainTransform/TrainTransformOBB, Mosaic center/randomization and RP policy, MixUp policy, OBB angle convention, training configuration, export behavior, repository layout, CI, and artifacts.\n\nRefs #65

**Comment by shapovalovei** [comment]
Orchestrator implementation review accepted PR #66 on exact head `0ee6ef9c49264ac7439801b0cb1eb3a4fc9141bb`.

Independent review confirmed the production change is exactly the two-line causal correction authorized by #65: enabled `MosaicDetection` and `MosaicDetectionOBB` use transient active wrapper `self.input_dim` instead of stale `self._dataset.input_dim`. The focused maintained-path regressions cover square/rectangular transitions, repeated changes, random-perspective canvas/border/output dimensions, target coordinate space, Mosaic phase boundaries, no-Mosaic preservation, decorator state cleanup, and MixUp dimension forwarding. Broader local suite passed.

No implementation blockers found. Merge is temporarily blocked only because PR #66 remains Draft and the connected GitHub Draft→Ready mutation fails on the known GraphQL `fullDatabaseId` connector error. Owner should mark PR #66 Ready for review manually; orchestrator must then re-check exact head before squash merge.

**Comment by shapovalovei** [comment]
Orchestrator post-merge completion: PR #66 was independently reviewed and squash-merged from exact reviewed head `0ee6ef9c49264ac7439801b0cb1eb3a4fc9141bb`. GitHub merge SHA: `3b3c33413fb29481f3655129eac090abb3800896`. The implementation stayed within the two Mosaic wrapper dimension-source lines plus focused regression coverage; no sampler/head/transform/training/export scope expansion was accepted. Closing #65 as Completed.
